src/zarr/registry.py

```diff
@@ -0,0 +1,144 @@
+from __future__ import annotations
+
+import warnings
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from zarr.abc.codec import Codec, CodecPipeline
+    from zarr.buffer import Buffer, NDBuffer
+
+from importlib.metadata import EntryPoint
+from importlib.metadata import entry_points as get_entry_points
+
+from zarr.config import BadConfigError, config
+
+__codec_registry: dict[str, dict[str, type[Codec]]] = {}
+__lazy_load_codecs: dict[str, EntryPoint] = {}
+__pipeline_registry: dict[str, type[CodecPipeline]] = {}
+__lazy_load_pipelines: list[EntryPoint] = []
+__buffer_registry: dict[str, type[Buffer]] = {}
+__lazy_load_buffer: list[EntryPoint] = []
+__ndbuffer_registry: dict[str, type[NDBuffer]] = {}
+__lazy_load_ndbuffer: list[EntryPoint] = []
+
+"""
+The registry module is responsible for managing implementations of codecs, pipelines, buffers and ndbuffers and 
+collecting them from entrypoints.
+The implementation used is determined by the config
+"""
+
+
+def _collect_entrypoints() -> (
+    tuple[dict[str, EntryPoint], list[EntryPoint], list[EntryPoint], list[EntryPoint]]
+):
+    """Collects codecs, pipelines, buffers and ndbuffers from entrypoints"""
+    entry_points = get_entry_points()
+    for e in entry_points.select(group="zarr.codecs"):
+        __lazy_load_codecs[e.name] = e
+    for e in entry_points.select(group="zarr"):
```

Comment:
What happens in multiple libraries in my env declare a `buffer` class? How will I be able to pick the right one?

Reply:
all will be registered and the correct one according to `config` will be chosen in `get_buffer_class`. With using the fully qualified name this should be an unambiguous assignment

Reply:
do you mean multiple implementations per entrypoint?

Reply:
oh, multiple implementations of the same codec would overwrite each other in the __lazy_load_codecs dict

Reply:
I think that needs to change. The idea is that multiple libraries can provide implementations for a specific codec (e.g. CPU- and GPU-based gzip) via entrypoints and the user can select one of these implementations via config.

I was also concerned about the other entrypoints (e.g. buffer, pipeline). It would be nice if libraries could declare more than one of each. Again, the user could then select the class via the config through the fully-qualified name.

Reply:
hmm how would one specify that in `entry_points.txt`?
my ideas were 
1. seperate different implementations comma-separated like
```
[zarr.codecs]
gzip = package:EntrypointGzipCodec1, package:EntrypointGzipCodec2
[zarr]
buffer = package:TestEntrypointBuffer, package:AnotherTestEntrypointBuffer
```
but that would deviate from the PEP 508 standard of entry_points.txt 

2. provide names for different implementations (which do not have an effect) and have a group for buffer, ndbuffer, pipeline and every Codec 
```
[zarr.codecs.gzip]
some_name = package:EntrypointGzipCodec1
another = package:EntrypointGzipCodec2
[zarr.buffer]
some_name = package:TestEntrypointBuffer
another = package:AnotherTestEntrypointBuffer
```
I'd prefer the second but I dont like having random names for each implementation that will not be used at parsing
any other ideas?

Reply:
Would it be possible to support both single items (e.g. `zarr.buffer = package:TestEntrypointBuffer`) and groups (e.g. `zarr.buffer = { gpu_buffer = package:TestEntrypointBuffer, cpu_buffer = package:AnotherTestEntrypointBuffer }`)?

Reply:
good idea! allowed syntax is now all of the following:
```
[zarr.codecs]
gzip = package:EntrypointGzipCodec1
[zarr.codecs.gzip]
some_name = package:EntrypointGzipCodec2
another = package:EntrypointGzipCodec3

[zarr]
buffer = package:TestBuffer1
[zarr.buffer]
xyz = package:TestBuffer2
abc = package:TestBuffer3
```